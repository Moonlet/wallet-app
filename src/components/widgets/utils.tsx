import React from 'react';
import { DataType, ICurrencyData, IData, IDataStyle, ITextData } from './types';
import { Text } from '../../library';
import BigNumber from 'bignumber.js';
import { formatNumber } from '../../core/utils/format-number';
import { normalize, normalizeFontAndLineHeight } from '../../styles/dimensions';

const formatTextData = (data: ITextData): string => {
    let text = '';
    // Static text
    text = String(data.value);

    if (data?.params) {
        // TODO
    }

    // Translated text
    // TODO

    return text;
};

const amountFromStd = (value: BigNumber | number | string, decimals: number): BigNumber => {
    return new BigNumber(value).dividedBy(new BigNumber(10).pow(decimals));
};

const formatCurrencyData = (data: ICurrencyData): string => {
    let text = '';
    let value = data.value;

    if (data?.decimals) {
        value = amountFromStd(value, data.decimals).toFixed();
    }

    text = formatNumber(Number(value), {
        currency: data.symbol,
        maximumFractionDigits: data?.round?.decimals
    });

    if (data?.beautify) {
        const out = beautify(value);
        text = new BigNumber(out.value).toFixed(data.beautify.decimals) + out.unit;
    }

    return text;
};

const formatStyles = (style: IDataStyle): { [key: string]: string | number } => {
    const finalStyle = {};

    for (const s of Object.keys(style)) {
        const sVal: any = style[s];

        if (typeof sVal === 'object') {
            if (sVal.fn === 'normalize') {
                finalStyle[s] = normalize(sVal.value);
            }
            if (sVal.fn === 'normalizeFontAndLineHeight') {
                finalStyle[s] = normalizeFontAndLineHeight(sVal.value);
            }
        } else {
            finalStyle[s] = style[s];
        }
    }

    return finalStyle;
};

const beautify = (
    value: BigNumber | string | number,
    fromValue?: number
): {
    value: BigNumber | string;
    unit: string;
} => {
    fromValue = fromValue || 1000;

    value = new BigNumber(value);

    let out = { value, unit: '' };

    if (value.gte(fromValue)) {
        const units = [
            [1000000000, 'B'],
            [1000000, 'M'],
            [1000, 'k']
        ];

        for (const unit of units) {
            if (value.gte(unit[0])) {
                out = {
                    value: value.div(unit[0]),
                    unit: unit[1].toString()
                };
                break;
            }
        }
    }

    return out;
};

export const formatDataJSXElements = (data: IData | IData[], baseStyle: any): JSX.Element[] => {
    if (!Array.isArray(data)) {
        data = [data];
    }

    const finalData: JSX.Element[] = [];

    for (const d of data) {
        switch (d.type) {
            case DataType.TEXT:
                finalData.push(
                    <Text style={[baseStyle, d?.style && formatStyles(d.style)]}>
                        {formatTextData(d.data as ITextData)}
                    </Text>
                );
                break;

            case DataType.CURRENCY:
                finalData.push(
                    <Text style={[baseStyle, d?.style && formatStyles(d.style)]}>
                        {formatCurrencyData(d.data as ICurrencyData)}
                    </Text>
                );
                break;

            default:
                break;
        }
    }

    return finalData;
};
