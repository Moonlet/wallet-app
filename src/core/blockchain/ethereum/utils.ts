import { TransactionStatus } from '../../wallet/types';

export const getEthStatusByCode = (status): TransactionStatus => {
    switch (parseInt(status, 16)) {
        case 0:
            return TransactionStatus.FAILED;
        case 1:
            return TransactionStatus.SUCCESS;
        case 2:
            return TransactionStatus.PENDING;
        default:
            return TransactionStatus.FAILED;
    }
};
