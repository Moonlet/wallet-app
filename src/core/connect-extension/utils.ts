import { getBlockchain } from '../blockchain/blockchain-factory';
import { NavigationParams } from 'react-navigation';
import { NavigationService } from '../../navigation/navigation-service';

export const getUrlParams = (search = ``) => {
    const hashes = search.slice(search.indexOf('?') + 1).split('&');
    return hashes.reduce((params, hash) => {
        const split = hash.indexOf('=');

        if (split < 0) {
            return { ...params, [hash]: null };
        }

        const key = hash.slice(0, split);
        const val = hash.slice(split + 1);

        return { ...params, [key]: decodeURIComponent(val) };
    }, {});
};

export const signExtensionTransaction = payload => {
    const data = payload.params[0];
    const blockchainConfig = getBlockchain(data.selectedAccount.blockchain).config;
    const navigationParams: NavigationParams = {
        blockchain: data.selectedAccount.blockchain,
        accountIndex: data.selectedAccount.index,
        token: data.token,
        tokenLogo: blockchainConfig.tokens[blockchainConfig.coin].icon,
        extensionTransactionPayload: payload
    };

    if (NavigationService.getCurrentRoute() === 'Token') {
        NavigationService.replace('Token', navigationParams);
    } else {
        NavigationService.navigate('Token', navigationParams);
    }
};
