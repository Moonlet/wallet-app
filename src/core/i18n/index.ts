export * from './translation/translate-component';
export * from './translation/translate';
