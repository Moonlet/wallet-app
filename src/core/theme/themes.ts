export enum Theme {
    dark = 'dark',
    light = 'light'
}
