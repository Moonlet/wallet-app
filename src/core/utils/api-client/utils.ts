import moment from 'moment';
import CONFIG from '../../../config';
import * as schnorr from '@zilliqa-js/crypto/dist/schnorr';

// Refactor this using ntp time
export const getTimestamp = () => {
    return moment().unix() * 1000;
};

export const getDomain = () => {
    return new URL(CONFIG.walletApiBaseUrl).host;
};

export const getSignature = (
    data: any,
    walletPrivateKey: string,
    walletPublicKey: string
): string => {
    const sig = schnorr.sign(
        Buffer.from(JSON.stringify(data), 'hex'),
        Buffer.from(walletPrivateKey, 'hex'),
        Buffer.from(walletPublicKey, 'hex')
    );

    return sig.r.toString('hex') + sig.s.toString('hex');
};
