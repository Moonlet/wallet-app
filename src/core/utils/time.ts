export const delay = (time: number = 0) =>
    new Promise(resolver => setTimeout(() => resolver(), time));
