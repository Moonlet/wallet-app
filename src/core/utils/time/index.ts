export const timeUtils = {
    delay: (time = 0) => new Promise(resolver => setTimeout(() => resolver(), time))
};
