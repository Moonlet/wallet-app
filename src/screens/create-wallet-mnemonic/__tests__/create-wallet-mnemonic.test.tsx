import React from 'react';
import {
    CreateWalletMnemonicScreenComponent,
    IProps,
    navigationOptions,
    IReduxProps
} from '../create-wallet-mnemonic';
import { darkTheme } from '../../../styles/themes/dark-theme';
import styleProvider from '../styles';

import { shallow } from 'enzyme';
import { loadTranslations } from '../../../core/i18n';
import { Mnemonic } from '../../../core/wallet/hd-wallet/mnemonic';
import { timeUtils } from '../../../core/utils/time';

const props: IProps & IReduxProps = {
    // @ts-ignore
    navigation: {
        navigate: jest.fn()
    },
    styles: styleProvider(darkTheme),
    tosVersion: 1
};

jest.mock('../../../core/wallet/hd-wallet/mnemonic');
jest.mock('../../../core/constants/app', () => {
    return {
        TOS_VERSION: 1
    };
});

describe('creat wallet terms screen component', () => {
    beforeAll(async () => {
        Mnemonic.generate = jest.fn(() =>
            Promise.resolve('1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24')
        );

        await loadTranslations('en');
    });

    it('renders correctly', async () => {
        const wrapper = shallow(<CreateWalletMnemonicScreenComponent {...props} />);
        await timeUtils.delay();
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('navigates to next screen', () => {
        const wrapper = shallow(<CreateWalletMnemonicScreenComponent {...props} />);
        wrapper.find('[testID="button-next"]').simulate('Press');
        expect(props.navigation.navigate).toHaveBeenCalledTimes(1);
    });

    it('sets correct navigation options', () => {
        const navigationProp = { navigation: { state: { params: { goBack: jest.fn() } } } };
        const options = navigationOptions(navigationProp);
        expect(options.headerLeft()).toMatchSnapshot();
        expect(options).toMatchSnapshot();
    });

    it('does not have a back button if no goBack param is set', () => {
        const navigationProp = { navigation: {} };
        const options = navigationOptions(navigationProp);
        expect(options.headerLeft()).toBe(null);
    });

    it('navigates to tos when there are new tos', () => {
        shallow(<CreateWalletMnemonicScreenComponent {...{ ...props, tosVersion: 0 }} />);
        expect(props.navigation.navigate).toHaveBeenCalledWith('CreateWalletTerms');
    });
});
