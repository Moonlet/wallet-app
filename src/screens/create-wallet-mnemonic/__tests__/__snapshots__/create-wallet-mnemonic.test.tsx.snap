// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creat wallet terms screen component renders correctly 1`] = `
"<View style={{...}}>
  <View style={{...}}>
    <View style={{...}}>
      <View style={{...}}>
        <withTheme(TextComponent) style={{...}}>
          1. 1
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          2. 2
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          3. 3
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          4. 4
        </withTheme(TextComponent)>
      </View>
      <View style={{...}}>
        <withTheme(TextComponent) style={{...}}>
          5. 5
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          6. 6
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          7. 7
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          8. 8
        </withTheme(TextComponent)>
      </View>
      <View style={{...}}>
        <withTheme(TextComponent) style={{...}}>
          9. 9
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          10. 10
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          11. 11
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          12. 12
        </withTheme(TextComponent)>
      </View>
      <View style={{...}}>
        <withTheme(TextComponent) style={{...}}>
          13. 13
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          14. 14
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          15. 15
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          16. 16
        </withTheme(TextComponent)>
      </View>
      <View style={{...}}>
        <withTheme(TextComponent) style={{...}}>
          17. 17
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          18. 18
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          19. 19
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          20. 20
        </withTheme(TextComponent)>
      </View>
      <View style={{...}}>
        <withTheme(TextComponent) style={{...}}>
          21. 21
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          22. 22
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          23. 23
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}}>
          24. 24
        </withTheme(TextComponent)>
      </View>
    </View>
    <View style={{...}}>
      <Icon name=\\"warning\\" size={24} style={{...}} allowFontScaling={false} />
      <withTheme(TextComponent) darker={true}>
        Please save the recovery phrase and keep it in a safe place. The recovery phrase is the only way to restore your Moonlet wallet in case you lose your phone or forget your password.
      </withTheme(TextComponent)>
    </View>
  </View>
  <View style={{...}}>
    <withTheme(ButtonComponent) style={{...}} onPress={[Function: onPress]}>
      Copy
    </withTheme(ButtonComponent)>
    <withTheme(ButtonComponent) testID=\\"button-next\\" primary={true} onPress={[Function: onPress]}>
      Next
    </withTheme(ButtonComponent)>
  </View>
</View>"
`;

exports[`creat wallet terms screen component sets correct navigation options 1`] = `
<withTheme(HeaderLeftComponent)
  icon="arrow-left-1"
  onPress={[Function]}
/>
`;

exports[`creat wallet terms screen component sets correct navigation options 2`] = `
Object {
  "headerLeft": [Function],
  "title": "Create",
}
`;
